programs/mango-v4/src/health/cache.rs

```diff
@@ -89,6 +89,51 @@ pub fn compute_health(
     Ok(new_health_cache(account, retriever)?.health(health_type))
 }
 
+/// How much of a token can be taken away before health decreases to zero?
+pub fn spot_amount_taken_for_health_zero(
+    mut health: I80F48,
+    starting_spot: I80F48,
+    asset_weighted_price: I80F48,
+    liab_weighted_price: I80F48,
+) -> Result<I80F48> {
+    if health <= 0 {
+        return Ok(I80F48::ZERO);
+    }
+
+    let mut taken_spot = I80F48::ZERO;
+    if starting_spot > 0 {
+        if asset_weighted_price > 0 {
+            let asset_max = health / asset_weighted_price;
+            if asset_max <= starting_spot {
+                return Ok(asset_max);
+            }
+        }
+        taken_spot = starting_spot;
+        health -= starting_spot * asset_weighted_price;
+    }
+    if health > 0 {
+        require_gt!(liab_weighted_price, 0);
+        taken_spot += health / liab_weighted_price;
+    }
+    Ok(taken_spot)
+}
+
+/// How much of a token can be gained before health increases to zero?
+pub fn spot_amount_given_for_health_zero(
+    health: I80F48,
+    starting_spot: I80F48,
+    asset_weighted_price: I80F48,
+    liab_weighted_price: I80F48,
+) -> Result<I80F48> {
+    // asset/liab prices are reversed intentionally
+    spot_amount_taken_for_health_zero(
+        -health,
+        -starting_spot,
```

Comment:
nit: can you please explain to me why this (negating the current position balance) works?

Reply:
Increasing health by depositing spot is equivalent to `spot_amount_taken_for_health_zero()`. You want to give spot and the weights change depending on the balance. 
- If spot balance is negative, use the liab weight. (equivalent to spot_amount_taken_for_health_zero() taking a positive balance with the asset weight)
- If spot balance is positive, use the asset weight. (equivalent to the other function taking a negative balance with the liab weight)